smarts/core/road_map.py

```diff
@@ -0,0 +1,437 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+# to allow for typing to refer to class being defined (RoadMap)
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import NamedTuple, List, Sequence, Tuple
+
+import numpy as np
+from shapely.geometry import Polygon
+
+from .coordinates import BoundingBox, Heading, Point, Pose, RefLinePoint
+from .utils.math import (
+    fast_quaternion_from_angle,
+    min_angles_difference_signed,
+    signed_dist_to_line,
+    vec_to_radians,
+)
+
+
+# TODO:
+# - also consider Esri, QGIS and Google Maps formats
+# - look at how OpenDrive encodes lane direction
+#    -https://www.asam.net/index.php?eID=dumpFile&t=f&f=4089&token=deea5d707e2d0edeeb4fccd544a973de4bc46a09
+
+
+class RoadMap:
+    """Base class from which map implementation classes extend."""
+
+    @property
+    def source(self) -> str:
+        raise NotImplementedError()
+
+    @property
+    def bounding_box(self) -> BoundingBox:
+        # maps are assumed to start at the origin
+        raise NotImplementedError()
+
+    @property
+    def scale_factor(self) -> float:
+        # map units per meter
+        return 1.0
+
+    def to_glb(self, at_path):
+        """ build a glb file for camera rendering and envision """
+        raise NotImplementedError()
+
+    def lane_by_id(self, lane_id: str) -> RoadMap.Lane:
+        raise NotImplementedError()
+
+    def road_by_id(self, road_id: str) -> RoadMap.Road:
+        raise NotImplementedError()
+
+    def nearest_lanes(
+        self, point: Point, radius: float = None, include_junctions=True
+    ) -> List[Tuple[RoadMap.Lane, float]]:
+        raise NotImplementedError()
+
+    def nearest_lane(
+        self, point: Point, radius: float = None, include_junctions=True
+    ) -> RoadMap.Lane:
+        nearest_lanes = self.nearest_lanes(point, radius, include_junctions)
+        return nearest_lanes[0][0] if nearest_lanes else None
+
+    def road_with_point(self, point: Point) -> RoadMap.Road:
+        raise NotImplementedError()
```

Comment:
This is a very tough question (which is also related to my unfinished implementation of `foes` elsewhere).  I confess that I haven't worked out what to do here yet, and this is probably the biggest and most important outstanding "open question" remaining in the refactor.

You might remember that the original design of `RoadMap` had a `RoadMap.Junction` class in it at the same level as `RoadMap.Road`.   I recently took it out because I thought we could get away without it, assuming we somehow solve this "overlapping" roads problem in the implementation.  But I'm torn about this, and inclined to add it back in recently.  

There's a lot to write here.  This probably warrants an offline "design discussion" with the team, actually.  Will bring it up in our next meeting.

Reply:
By the way, I just added a note to the PR description about not digging too deeply on things related to this as I'm actively working on it.

Reply:
Note: in the current plan, there will only ever be one road hit by a point.  (It may have `is_junction` set to true though and contain lanes that cross.)

Reply:
So I've now updated the direction-related and intersection-related APIs to what I think their "final form" should be and removed the note from the PR description.